src/rt-app.c

```diff
@@ -1250,7 +1250,24 @@ int main(int argc, char* argv[])
 
 	/* Start the use case */
 	for (i = 0; i < nthreads; i++) {
-		tdata = &opts.threads_data[i];
+		/*
+		 * Duplicate thread data so that we can safely copy the
+		 * original thread data when forking.
+		 *
+		 * If we don't do that and try to fork a running thread, we
+		 * might have partially modified content of thread data since
+		 * thread_body() calls functions like set_thread_affinity()
+		 * that modifies thread data at runtime. Rather than introduce
+		 * complex locking ensure that the original parsed thread_data
+		 * are intact and duplicate them before forking or when we run
+		 * for the first time as in here.
+		 */
+		tdata = malloc(sizeof(thread_data_t));
+		if (!tdata) {
+			perror("Failed to duplicate thread data");
+			exit(EXIT_FAILURE);
+		}
+		memcpy(tdata, &opts.threads_data[i], sizeof(*tdata));
```

Comment:
Don't we need to free this up before thread exit as we do for forked ones?

Reply:
Good catch. I think we can free data unconditionally in thread_body() since I changed the logic to always malloc() tdata for forked and unforked tasks.